pkg/query-service/model/v3/v3.go

```diff
@@ -1142,3 +1142,92 @@ type MetricMetadataResponse struct {
 	IsMonotonic bool      `json:"isMonotonic"`
 	Temporality string    `json:"temporality"`
 }
+
+type LabelsString string
+
+func (l *LabelsString) MarshalJSON() ([]byte, error) {
+	lbls := make(map[string]string)
+	err := json.Unmarshal([]byte(*l), &lbls)
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(lbls)
+}
+
+func (l *LabelsString) Scan(src interface{}) error {
+	if data, ok := src.(string); ok {
+		*l = LabelsString(data)
+	}
+	return nil
+}
+
+func (l LabelsString) String() string {
+	return string(l)
+}
+
+type RuleStateHistory struct {
+	RuleID   string `json:"ruleID" ch:"rule_id"`
+	RuleName string `json:"ruleName" ch:"rule_name"`
+	// One of ["normal", "firing"]
+	OverallState        string `json:"overallState" ch:"overall_state"`
+	OverallStateChanged bool   `json:"overallStateChanged" ch:"overall_state_changed"`
+	// One of ["normal", "firing", "no_data", "muted"]
+	State        string       `json:"state" ch:"state"`
+	StateChanged bool         `json:"stateChanged" ch:"state_changed"`
+	UnixMilli    int64        `json:"unixMilli" ch:"unix_milli"`
+	Labels       LabelsString `json:"labels" ch:"labels"`
+	Fingerprint  uint64       `json:"fingerprint" ch:"fingerprint"`
+	Value        float64      `json:"value" ch:"value"`
+}
+
+type QueryRuleStateHistory struct {
+	Start   int64      `json:"start"`
+	End     int64      `json:"end"`
+	Filters *FilterSet `json:"filters"`
+	Offset  int64      `json:"offset"`
+	Limit   int64      `json:"limit"`
+	Order   string     `json:"order"`
+}
+
+func (r *QueryRuleStateHistory) Validate() error {
+	if r.Start == 0 || r.End == 0 {
+		return fmt.Errorf("start and end are required")
+	}
+	if r.Offset < 0 || r.Limit < 0 {
+		return fmt.Errorf("offset and limit must be greater than 0")
```

Comment:
set default limit if not provided?

Reply:
I think the client should consciously make decisions about how much it wants to consume so I am not adding any default.